### PR TITLE
extraction fix

### PR DESCRIPTION
**Pull request by kmdeck**
## Purpose 
Update root extraction so that it tends to zero as the soil dries or freezes.
We have the Darcy flux:
Flux = -K_eff [(psi_canopy - psi_soil)/(z_canopy -z_soil) + 1]

Previously: K_eff = [K_canopy(psi_soil) + K_canopy(psi_canopy)]/2.

Two things have changed: we now use the harmonic mean, and we replace K_canopy(psi_soil) with K_soil(psi_soil). Note that K_soil() and K_canopy() are two different functions. 

Currently: K_eff = K_soil(psi_soil) * K_canopy(psi_canopy)/ [K_soil(psi_soil) + K_canopy(psi_canopy)]

So, if K_soil(psi_soil) -> 0, then K_eff -> 0, and the root extraction goes to zero consistently. Although as psi_soil ->0, K_canopy(psi_soil) ->0, it may not drop to zero quickly enough due to the differing functional forms.

Please see CLM:
Screenshot 2024-09-10 at 15 24 54
which is different in that it uses conductances and has canopy -> root, but see Equation 11.16 (harmonic mean) and Equation 11.18 (K_soil in soil layer, i.e. K_soil(psi_soil))

## Effect on NaNs:
Global run after soil evaporation fix (very similar to current main):
swc_2 88576e7_new
Global run after these changes:
image

## To-do
Review

## Content
1. Change the root extraction to depend on K_soil, and not K_plant(psi_soil). Note that this is only changed in the prognostic soil +canopy mode, and not the standalone canopy mode (prescribed soil). In the latter, we use K_plant(psi_soil) still.
2. Use the geometric mean to compute K_eff, and not arithmetic mean
3. Update tests

----
- [ ] I have read and checked the items on the review checklist.

### CODE REVIEW

File: .buildkite/longruns_gpu/pipeline.yml [hunk 1]
```diff
@@ -37,7 +37,7 @@ steps:
         artifact_paths: "land_longrun_gpu/*png"
         agents:
           slurm_gpus: 1
-          slurm_time: 03:00:00
+          slurm_time: 08:00:00
```
Comment (juliasloan25) [inline .buildkite/longruns_gpu/pipeline.yml:40]: do we need this change? it looks like the run is still taking ~2.5 hours to run

**Comment by juliasloan25** [comment]
> Change the root extraction to depend on K_soil, and not K_plant(psi_soil). Note that this is only changed in the prognostic soil +canopy mode, and not the standalone canopy mode (prescribed soil). In the latter, we use K_plant(psi_soil) still.

It seems unusual to me to use slightly different equations for standalone vs integrated modes. Although maybe it doesn't make sense to evaluate K_soil with prescribed soil?

File: src/standalone/Vegetation/PlantHydraulics.jl [hunk 2]
```diff
@@ -342,13 +342,15 @@ Computes the water flux given the absolute potential (pressure/(ρg))
...
 the flow path between these two points.
 
-We currently assuming an arithmetic
-mean for mean K_sat between the two points (Bonan, 2019; Zhu, 2008)
-to take into account the change in K_sat halfway
-between z1 and z2; this is incorrect for compartments of differing sizes.
-"""
-function flux(z1, z2, ψ1, ψ2, K1, K2)
-    flux = -(K1 + K2) / 2 * ((ψ2 - ψ1) / (z2 - z1) + 1)
+We currently assuming an geometric
```
Comment (juliasloan25) [inline src/standalone/Vegetation/PlantHydraulics.jl:345]:
```suggestion
We are currently asuuming a geometric
```

Comment (juliasloan25) [reply src/standalone/Vegetation/PlantHydraulics.jl:345]: it would be nice to also cite CLM here

File: src/standalone/Vegetation/PlantHydraulics.jl [hunk 3]
```diff
@@ -342,13 +342,15 @@ Computes the water flux given the absolute potential (pressure/(ρg))
...
+mean for mean K_sat between the two points. Previously,
+we used the arithmetic mean (Bonan, 2019; Zhu, 2008),
+but then when the soil K was very low, root extraction would
+continue. This should be modified for compartments of differing sizes.
```
Comment (juliasloan25) [inline src/standalone/Vegetation/PlantHydraulics.jl:349]: what do you mean by "This should be modified for compartments of differing sizes."? what should be modified?

File: src/standalone/Vegetation/PlantHydraulics.jl [hunk 4]
```diff
@@ -342,13 +342,15 @@ Computes the water flux given the absolute potential (pressure/(ρg))
...
+"""
+function flux(z1::FT, z2::FT, ψ1::FT, ψ2::FT, K1::FT, K2::FT) where {FT}
```
Comment (juliasloan25) [inline src/standalone/Vegetation/PlantHydraulics.jl:351]: I think a more descriptive name than "flux" would be nice - maybe "water_flux"? up to you

File: test/standalone/Vegetation/plant_hydraulics_test.jl [hunk 5]
```diff
@@ -215,7 +215,7 @@ for FT in (Float32, Float64)
             SAI,
             RAI,
         )
-        K_sat_plant = 1.8e-8 # m/s.
+        K_sat_plant = 1.0 # m/s.
```
Comment (juliasloan25) [inline test/standalone/Vegetation/plant_hydraulics_test.jl:218]: why did you change this value (and T, LAI further down) to 1?

File: test/standalone/Vegetation/plant_hydraulics_test.jl [hunk 6]
```diff
@@ -307,7 +307,7 @@ for FT in (Float32, Float64)
                                     vcat(root_depths, [0.0])[2:end] -
                                     vcat(root_depths, [0.0])[1:(end - 1)]
                                 ),
-                            ) * AI[:root]
+                            ) * AI[:stem]
```
Comment (juliasloan25) [inline test/standalone/Vegetation/plant_hydraulics_test.jl:310]: why did this change from root to stem?

**Review by juliasloan25 — COMMENTED** [review]
Awesome!! This looks great, I just had a couple comments :)

**Comment by kmdeck** [comment]
> > Change the root extraction to depend on K_soil, and not K_plant(psi_soil). Note that this is only changed in the prognostic soil +canopy mode, and not the standalone canopy mode (prescribed soil). In the latter, we use K_plant(psi_soil) still.
> 
> It seems unusual to me to use slightly different equations for standalone vs integrated modes. Although maybe it doesn't make sense to evaluate K_soil with prescribed soil?

Yeah, I agree...it would entail changing the PrescribedSoil struct to include a function for K_soil(psi) as well, or a timeseries of Ksoil(t) which is consistent with psi_soil(t), which would be made up at this point. We dont really have a use case for it yet.

File: .buildkite/longruns_gpu/pipeline.yml (hunk 1 above)
Comment (kmdeck) [reply .buildkite/longruns_gpu/pipeline.yml:40]: thanks!

File: experiments/long_runs/land.jl [hunk 7]
```diff
@@ -628,7 +629,7 @@ end
 function setup_and_solve_problem(; greet = false)
 
     t0 = 0.0
-    tf = 60 * 60.0 * 24 * 14
+    tf = 60 * 60.0 * 24 * 340
```
Comment (braghiere) [inline experiments/long_runs/land.jl:632]: Any reason why you stop in 340? We have a full 365 days right?

File: src/standalone/Vegetation/PlantHydraulics.jl (hunk 4 above)
Comment (braghiere) [reply src/standalone/Vegetation/PlantHydraulics.jl:355]: +1

File: test/standalone/Vegetation/plant_hydraulics_test.jl [hunk 8]
```diff
@@ -284,8 +284,8 @@ for FT in (Float32, Float64)
             )
             # Set system to hydrostatic equilibrium state by setting fluxes to zero, and setting LHS of both ODEs to 0
             function initial_compute_exp_tendency!(F, Y)
-                AI = (; leaf = LAI(0.0), root = RAI, stem = SAI)
-                T0A = FT(1e-8) * AI[:leaf]
+                AI = (; leaf = LAI(1.0), root = RAI, stem = SAI)
```
Comment (braghiere) [inline test/standalone/Vegetation/plant_hydraulics_test.jl:287]: Shouldn't this LAI vary spatially too? Why is it 1.0 everywhere?

File: test/standalone/Vegetation/plant_hydraulics_test.jl [hunk 9]
```diff
@@ -335,7 +332,9 @@ for FT in (Float32, Float64)
             soln = nlsolve(
                 initial_compute_exp_tendency!,
                 Vector{FT}(-0.03:0.01:0.07);
-                ftol = 1e-11,
+                ftol = eps(FT),
+                method = :newton,
+                iterations = 20,
```
Comment (braghiere) [inline test/standalone/Vegetation/plant_hydraulics_test.jl:337]: I am not sure this is relevant but are 20 iterations enough to converge everywhere? Perhaps some places would take more iterations?

File: src/standalone/Vegetation/PlantHydraulics.jl (hunk 2 above)
Comment (braghiere) [reply src/standalone/Vegetation/PlantHydraulics.jl]: +1

File: src/standalone/Vegetation/PlantHydraulics.jl (hunk 3 above)
Comment (kmdeck) [reply src/standalone/Vegetation/PlantHydraulics.jl]: I added a comment for what CLM does, but will double check.

**Comment by braghiere** [comment]
> > > Change the root extraction to depend on K_soil, and not K_plant(psi_soil). Note that this is only changed in the prognostic soil +canopy mode, and not the standalone canopy mode (prescribed soil). In the latter, we use K_plant(psi_soil) still.
> > 
> > 
> > It seems unusual to me to use slightly different equations for standalone vs integrated modes. Although maybe it doesn't make sense to evaluate K_soil with prescribed soil?
> 
> Yeah, I agree...it would entail changing the PrescribedSoil struct to include a function for K_soil(psi) as well, or a timeseries of Ksoil(t) which is consistent with psi_soil(t), which would be made up at this point. We dont really have a use case for it yet.

Agree. If it is not too difficult, maybe we can change the standalone as well? If it is a bit of work, we can leave it for now since we're focusing on canopy+soil development, which is more aligned to OKRs.

**Review by braghiere — COMMENTED** [review]
It looks good to me. Left a few comments.

File: src/standalone/Vegetation/PlantHydraulics.jl (hunk 4 above)
Comment (kmdeck) [reply src/standalone/Vegetation/PlantHydraulics.jl:355]: I changed this

File: src/standalone/Vegetation/PlantHydraulics.jl (hunk 3 above)
Comment (kmdeck) [reply src/standalone/Vegetation/PlantHydraulics.jl]: Ok, I fleshed out the comment more. This is something we probably should do. But it would require introducing additional parameters (root length density, root biomass). I believe the PlantHydraulics code will need some other improvements too, and I think we could tackle it then.

File: test/standalone/Vegetation/plant_hydraulics_test.jl (hunk 8 above)
Comment (kmdeck) [reply test/standalone/Vegetation/plant_hydraulics_test.jl:287]: In this case, it is just a test at a single location, so it is OK to have it be constant. but we vary spatially in our global and regional runs!

File: test/standalone/Vegetation/plant_hydraulics_test.jl (hunk 9 above)
Comment (kmdeck) [reply test/standalone/Vegetation/plant_hydraulics_test.jl:337]: good question! In this case, we are just running the test at one location/in a column, so it doenst matter, but yes, some places on the globe might take longer to converge. That said, I think many models still cap the number of iterations in places like this so that the entire run is not held up by a single point.

File: test/standalone/Vegetation/plant_hydraulics_test.jl (hunk 6 above)
Comment (kmdeck) [reply test/standalone/Vegetation/plant_hydraulics_test.jl:310]: In our src code, we use the above ground area index here, not the root area index. This was making sure the two align since this test is testing the src function.

File: test/standalone/Vegetation/plant_hydraulics_test.jl (hunk 5 above)
Comment (kmdeck) [reply test/standalone/Vegetation/plant_hydraulics_test.jl:218]: I changed it back - thought it turns out the LAI one was a typo! and our test must not have caught it because it was not strict enough

**Comment by juliasloan25** [comment]
> > > > Change the root extraction to depend on K_soil, and not K_plant(psi_soil). Note that this is only changed in the prognostic soil +canopy mode, and not the standalone canopy mode (prescribed soil). In the latter, we use K_plant(psi_soil) still.
> > > 
> > > 
> > > It seems unusual to me to use slightly different equations for standalone vs integrated modes. Although maybe it doesn't make sense to evaluate K_soil with prescribed soil?
> > 
> > 
> > Yeah, I agree...it would entail changing the PrescribedSoil struct to include a function for K_soil(psi) as well, or a timeseries of Ksoil(t) which is consistent with psi_soil(t), which would be made up at this point. We dont really have a use case for it yet.
> 
> Agree. If it is not too difficult, maybe we can change the standalone as well? If it is a bit of work, we can leave it for now since we're focusing on canopy+soil development, which is more aligned to OKRs.

I think it makes sense to keep PrescribedSoil as-is for now, until we do have a use case that necessitates this change

File: test/standalone/Vegetation/plant_hydraulics_test.jl (hunk 6 above)
Comment (juliasloan25) [reply test/standalone/Vegetation/plant_hydraulics_test.jl:310]: makes sense, thank you!

**Comment by kmdeck** [comment]
> > > > > Change the root extraction to depend on K_soil, and not K_plant(psi_soil). Note that this is only changed in the prognostic soil +canopy mode, and not the standalone canopy mode (prescribed soil). In the latter, we use K_plant(psi_soil) still.
> > > > 
> > > > 
> > > > It seems unusual to me to use slightly different equations for standalone vs integrated modes. Although maybe it doesn't make sense to evaluate K_soil with prescribed soil?
> > > 
> > > 
> > > Yeah, I agree...it would entail changing the PrescribedSoil struct to include a function for K_soil(psi) as well, or a timeseries of Ksoil(t) which is consistent with psi_soil(t), which would be made up at this point. We dont really have a use case for it yet.
> > 
> > 
> > Agree. If it is not too difficult, maybe we can change the standalone as well? If it is a bit of work, we can leave it for now since we're focusing on canopy+soil development, which is more aligned to OKRs.
> 
> I think it makes sense to keep PrescribedSoil as-is for now, until we do have a use case that necessitates this change

Thanks guys! I updated an old issue to have this included: https://github.com/CliMA/ClimaLand.jl/issues/159

**Comment by kmdeck** [comment]
@braghiere @juliasloan25 I will extend the run to 365 days (or document why we cant and open an issue, I forget why it didnt run before).

would this be ready to merge then? If so, if you approve it now Ill merge it after the last fix to set tf=365 days.

I would really like to rethink aspects of our plant hydraulics model thoroughly #159 , but for now want to move forward with what we have

**Comment by juliasloan25** [comment]
> @braghiere @juliasloan25 I will extend the run to 365 days (or document why we cant and open an issue, I forget why it didnt run before).
> 
> would this be ready to merge then? If so, if you approve it now Ill merge it after the last fix to set tf=365 days.
> 
> I would really like to rethink aspects of our plant hydraulics model thoroughly #159 , but for now want to move forward with what we have

Yeah, looks good to me!

**Review by juliasloan25 — APPROVED** [review]

File: experiments/long_runs/land.jl (hunk 7 above)
Comment (kmdeck) [reply experiments/long_runs/land.jl]: I extended this in both our long runs

**Review by braghiere — APPROVED** [review]
Looks good. Thanks!
